daemon/endpoint.go

```diff
@@ -211,19 +211,37 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 		}
 	}
 
+	if err := endpointmanager.AddEndpoint(d, ep, "Create endpoint from API PUT"); err != nil {
+		log.WithError(err).Warn("Aborting endpoint join")
+		return PutEndpointIDFailedCode, err
+	}
+
 	ep.UpdateLabels(d, addLabels, infoLabels, true)
 
-	// Do not add endpoint to the endpoint manager if client abort connection
 	select {
 	case <-ctx.Done():
 		log.WithError(ctx.Err()).Warn("Aborting endpoint join due client connection closed")
+		d.deleteEndpoint(ep)
 		return PutEndpointIDFailedCode, fmt.Errorf("aborting endpoint %d join due client connection closed", ep.ID)
 	default:
 	}
 
-	if err := endpointmanager.AddEndpoint(d, ep, "Create endpoint from API PUT"); err != nil {
-		log.WithError(err).Warn("Aborting endpoint join")
-		return PutEndpointIDFailedCode, err
+	if err := ep.LockAlive(); err != nil {
+		d.deleteEndpoint(ep)
+		return PutEndpointIDFailedCode, fmt.Errorf("endpoint was deleted while waiting for the identity")
+	}
+
+	build := ep.GetStateLocked() == endpoint.StateReady
```

Comment:
I don't understand this line, why is the endpoint only being built if it's ready? AFAIK the CNI plugin sends the endpoint to the cilium-agent in the `endpoint.StateWaitingToRegenerate` state

Reply:
cc: @tgraf  - I cherry-picked this commit per his recommendation. @tgraf can you provide some context here? I can add it as a comment for future reference.

Reply:
This change is no longer part of this PR so marking as resolved.